components/brave_wallet/browser/asset_discovery_manager.cc

```diff
@@ -330,6 +365,248 @@ void AssetDiscoveryManager::MergeDiscoveredEthAssets(
+  if (!nfts || !nfts->is_list()) {
+    return absl::nullopt;
+  }
+
+  const base::flat_map<std::string, std::string>& chain_ids =
+      FromSimpleHashChainId();
+  std::vector<mojom::BlockchainTokenPtr> nft_tokens;
+  for (const auto& nft : nfts->GetList()) {
+    auto token = mojom::BlockchainToken::New();
+
+    // Skip all tokens with a collection.spam_score > 0.
+    auto* collection = nft.FindDictKey("collection");
+    if (!collection) {
+      continue;
+    }
+    absl::optional<int> spam_score = collection->FindIntKey("spam_score");
+    if (!spam_score || *spam_score > 0) {
+      continue;
+    }
+
+    // contract_address (required)
+    auto* contract_address = nft.FindStringKey("contract_address");
+    if (!contract_address) {
+      continue;
+    }
+    token->contract_address = *contract_address;
+
+    // name
+    auto* name = nft.FindStringKey("name");
+    if (name) {
+      token->name = *name;
+    }
+
+    // logo
+    auto* logo = nft.FindStringKey("image_url");
+    if (logo) {
+      token->logo = *logo;
+    }
+
+    // is_erc20
+    token->is_erc20 = false;
+
+    // The contract dict has the standard information
+    // so we skip if it's not there.
+    auto* contract = nft.FindDictKey("contract");
+    if (!contract) {
+      continue;
+    }
+    auto* type = contract->FindStringKey("type");
+    if (!type) {
+      continue;
+    }
+
+    // is_erc721
+    if (coin == mojom::CoinType::ETH) {
+      bool is_erc721 = base::EqualsCaseInsensitiveASCII(*type, "ERC721");
+      if (!is_erc721) {
+        continue;
+      }
+      token->is_erc721 = true;
+    } else {  // mojom::CoinType::SOL
+      // Solana NFTs must be "NonFungible"
+      if (!base::EqualsCaseInsensitiveASCII(*type, "NonFungible")) {
+        continue;
+      }
+      token->is_erc721 = false;
+    }
+
+    // is_nft
+    token->is_nft = true;
+
+    // symbol
+    auto* symbol = contract->FindStringKey("symbol");
+    if (symbol) {
+      token->symbol = *symbol;
+    }
+
+    // decimals
+    token->decimals = 0;
+
+    // visible
+    token->visible = true;
+
+    // token_id (required for ETH only)
+    if (coin == mojom::CoinType::ETH) {
+      auto* token_id = nft.FindStringKey("token_id");
+      if (!token_id) {
+        continue;
+      }
+      uint256_t token_id_uint256;
+      if (!Base10ValueToUint256(*token_id, &token_id_uint256)) {
+        continue;
+      }
+      token->token_id = Uint256ValueToHex(token_id_uint256);
+    }
+
+    // chain_id (required)
+    auto* chain = nft.FindStringKey("chain");
+    if (!chain || !chain_ids.contains(*chain)) {
+      continue;
+    }
+    token->chain_id = chain_ids.at(*chain);
+
+    // coin
+    token->coin = coin;
+
+    nft_tokens.push_back(std::move(token));
+  }
+
+  return std::make_pair(nextURL, std::move(nft_tokens));
+}
+
+void AssetDiscoveryManager::DiscoverNFTsOnAllSupportedChains(
+    std::map<mojom::CoinType, std::vector<std::string>>& account_addresses,
```

Comment:
Should be `const std::map<...>&`, same for AssetDiscoveryManager::DiscoverAssetsOnAllSupportedChainsRefresh.